flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/LegacyScheduleOrUpdateConsumersTest.java

```diff
@@ -0,0 +1,168 @@
+
+import static org.apache.flink.runtime.jobmanager.SlotCountExceedingParallelismTest.SubtaskIndexReceiver.CONFIG_KEY;
+
+public class LegacyScheduleOrUpdateConsumersTest extends TestLogger {
+
+	private static final int NUMBER_OF_TMS = 2;
+	private static final int NUMBER_OF_SLOTS_PER_TM = 2;
+	private static final int PARALLELISM = NUMBER_OF_TMS * NUMBER_OF_SLOTS_PER_TM;
+
+	private static TestingCluster flink;
+
+	@BeforeClass
+	public static void setUp() throws Exception {
+		flink = TestingUtils.startTestingCluster(
+				NUMBER_OF_SLOTS_PER_TM,
+				NUMBER_OF_TMS,
+				TestingUtils.DEFAULT_AKKA_ASK_TIMEOUT());
+	}
+
+	@AfterClass
+	public static void tearDown() throws Exception {
+		flink.stop();
+	}
+
+	/**
+	 * Tests notifications of multiple receivers when a task produces both a pipelined and blocking
+	 * result.
+	 *
+	 * <pre>
+	 *                             +----------+
+	 *            +-- pipelined -> | Receiver |
+	 * +--------+ |                +----------+
+	 * | Sender |-|
+	 * +--------+ |                +----------+
+	 *            +-- blocking --> | Receiver |
+	 *                             +----------+
+	 * </pre>
+	 *
+	 * The pipelined receiver gets deployed after the first buffer is available and the blocking
+	 * one after all subtasks are finished.
+	 */
+	@Test
+	public void testMixedPipelinedAndBlockingResults() throws Exception {
+		final JobVertex sender = new JobVertex("Sender");
+		sender.setInvokableClass(BinaryRoundRobinSubtaskIndexSender.class);
+		sender.getConfiguration().setInteger(BinaryRoundRobinSubtaskIndexSender.CONFIG_KEY, PARALLELISM);
+		sender.setParallelism(PARALLELISM);
+
+		final JobVertex pipelinedReceiver = new JobVertex("Pipelined Receiver");
+		pipelinedReceiver.setInvokableClass(SlotCountExceedingParallelismTest.SubtaskIndexReceiver.class);
+		pipelinedReceiver.getConfiguration().setInteger(CONFIG_KEY, PARALLELISM);
+		pipelinedReceiver.setParallelism(PARALLELISM);
+
+		pipelinedReceiver.connectNewDataSetAsInput(
+				sender,
+				DistributionPattern.ALL_TO_ALL,
+				ResultPartitionType.PIPELINED);
+
+		final JobVertex blockingReceiver = new JobVertex("Blocking Receiver");
+		blockingReceiver.setInvokableClass(SlotCountExceedingParallelismTest.SubtaskIndexReceiver.class);
+		blockingReceiver.getConfiguration().setInteger(CONFIG_KEY, PARALLELISM);
+		blockingReceiver.setParallelism(PARALLELISM);
+
+		blockingReceiver.connectNewDataSetAsInput(sender,
+				DistributionPattern.ALL_TO_ALL,
+				ResultPartitionType.BLOCKING);
+
+		SlotSharingGroup slotSharingGroup = new SlotSharingGroup(
+				sender.getID(), pipelinedReceiver.getID(), blockingReceiver.getID());
+
+		sender.setSlotSharingGroup(slotSharingGroup);
+		pipelinedReceiver.setSlotSharingGroup(slotSharingGroup);
+		blockingReceiver.setSlotSharingGroup(slotSharingGroup);
+
+		final JobGraph jobGraph = new JobGraph(
+				"Mixed pipelined and blocking result",
+				sender,
+				pipelinedReceiver,
+				blockingReceiver);
```

Comment:
Could we deduplicate this code?

Reply:
You mean sharing it across legacy/new versions?

Reply:
Yes, or is there a dependency on a static lock?

Reply:
AFAIK there is no dependency.

It's just that given that the duality of these tests is limited to 1.5 I wouldn't in general take extra steps to re-use code. We would either have to create abstract base classes or move parts of the tests into public/package-private methods, which then we have to revert again in 1.6 (at least in part).

I'm questioning whether that is worth it, besides the added risk of accidentally modifying the behavior of the old test.

Reply:
Alright, if there is no easy way, then let's avoid the extra work and concentrate on quickly removing it afterwards.